### PR TITLE
Document Services for Extensions

### PR DESCRIPTION
**Pull request by estheragbaje**
This PR includes documentation for various services.

### CODE REVIEW

File: docs/extensions/services/introduction.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+Various services are available within Directus, such as `ItemsService`, `UserService`, `RolesService`, and others. The
+majority of these services extend the `ItemsService`, providing a unified approach to manipulating data. However, there
+are also standalone services.
+
+::: tip List of Available Services
+
+Directus provides a number of built-in services. Check out the full list
+[here](https://github.com/directus/directus/tree/bbefc62ef4727edb0b25eaafb6bb44273f79f834/api/src/services).
+
+:::
```
Comment (br41nslug) [inline docs/extensions/services/introduction.md:36]: We should put more emphasis on the other services! The big example being someone accidentally using `ItemsService('directus_users'` instead of the `UsersService` causing the password to be saved in plain text instead of hashed.

File: docs/extensions/services/accessing-database.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    await service.deleteOne(id);
+    res.json();
+  });
+```
+
+::: tip Explore ItemsService In-depth
+
+Check out the full list
+[here](https://github.com/directus/directus/blob/bbefc62ef4727edb0b25eaafb6bb44273f79f834/api/src/services/items.ts).
```
Comment (Boegie19) [inline docs/extensions/services/accessing-database.md:71]: I think using a dynamic like like https://github.com/directus/directus/blob/main/api/src/services/items.ts would be better then a perm link here. since lets say we do an update to the items service this one is pind to commit bbefc62ef4727edb0b25eaafb6bb44273f79f834 so any changes that happend after that would not be seen. second option would be to make it a snipit with https://github.com/directus/directus/blob/bbefc62ef4727edb0b25eaafb6bb44273f79f834 and then every release change the hash to the latest version and the snipit would accept "/api/src/services/items.ts" as input

File: docs/extensions/services/accessing-database.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
+---
+description: Learn about the ItemsService in Directus and how to utilize them when building extensions.
+contributors: Esther Agbaje
+---
+
+# Accessing Items in the Database
```
Comment (phazonoverload) [inline docs/extensions/services/accessing-database.md:6]: rename as per config

File: docs/extensions/services/accessing-database.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+```js
+  router.post('/items/:collection', async (req, res) => {
+    const schema = await getSchema();
+    const { collection } = req.params;
+
+    const service = new ItemsService(collection, { schema });
+    const data = await service.createOne(req.body);
+
+    res.json(data);
+  });
```
Comment (phazonoverload) [inline docs/extensions/services/accessing-database.md:22]: Unindent

File: docs/extensions/services/accessing-database.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+description: Learn about the ItemsService in Directus and how to utilize them when building extensions.
+contributors: Esther Agbaje
+---
+
+# Accessing Items in the Database
+
+The `ItemsService` allows you to work with and retrieve items data by creating endpoints to items in a collection. It
+provides a way to perform CRUD operations on items in a database.
+
+## Create an Item in a Specific Collection
```
Comment (phazonoverload) [inline docs/extensions/services/accessing-database.md:11]:
```suggestion
## Create an Item
```

I would drop 'in a specific collection' across the board because they are all in a specific collection. Directus users should understand the fact items must be inserted in a collection

File: docs/extensions/services/accessing-database.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+```js
+router.get('/items/:collection/:id', async (req, res) => {
+    const schema = await getSchema();
+
+    const { collection, id } = req.params;
+    const service = new ItemsService(collection, { schema });
+
+    const data = await service.readOne(id);
+    res.json(data);
+  });
```
Comment (phazonoverload) [inline docs/extensions/services/accessing-database.md:36]: Fix indentation. Make sure it's consistent.

File: docs/extensions/services/accessing-database.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
+---
+description: Learn about the ItemsService in Directus and how to utilize them when building extensions.
+contributors: Esther Agbaje
+---
+
+# Accessing Items in the Database
+
+The `ItemsService` allows you to work with and retrieve items data by creating endpoints to items in a collection. It
+provides a way to perform CRUD operations on items in a database.
+
```
Comment (phazonoverload) [inline docs/extensions/services/accessing-database.md:10]: There's a bit missing here which is just talking about the initialization. It should cover the fact that the service requires a collection and a schema. And that you must import and use `getSchema()`.

```js
export default defineEndpoint((router, context) => {
  const { services, getSchema } = context;
  const { ItemsService } = services;

  router.get('/', async (req, res) => {
    const schema = await getSchema();
    const service = new ItemsService('collection_name', { schema });
  });
});
```

By doing this, the other examples are contextualized more. This is a simple example with a hardcoded collection and just a `/` route handler. For this top-of-file example, let's keep it as easy as possible. 

Questions:
1. Can `getSchema()` be called at the top level of the endpoint or must it be in the route handler?  Is there a standard best practice? If anything goes and there's no benefit to putting it in the route handler I'd move it up to remove repetition if there are multiple endpoints. 
2. What else can go in the options object other than schema? Are they worth mentioning? Accountability comes to mind?

File: docs/extensions/services/accessing-database.md (hunk 6 above)
Comment (phazonoverload) [reply docs/extensions/services/accessing-database.md:36]: General across the board so I'll stop adding this comment but please do a sweep.

File: docs/extensions/services/accessing-database.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    await service.deleteOne(id);
+    res.json();
+  });
+```
+
+::: tip Explore ItemsService In-depth
+
+Check out the full list [here](https://github.com/directus/directus/blob/main/api/src/services/items.ts).
+
+:::
```
Comment (phazonoverload) [inline docs/extensions/services/accessing-database.md:72]:
```suggestion
::: tip Explore ItemsService In-depth

[Check out the full list of methods here](https://github.com/directus/directus/blob/main/api/src/services/items.ts).

:::
```

Comment (phazonoverload) [reply docs/extensions/services/accessing-database.md:72]: Universal change - make sure the link text has the full context for screenreaders.

File: docs/extensions/services/interacting-with-files.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
+---
+description: Learn about the FilesService in Directus and how to utilize them when building extensions.
+contributors: Esther Agbaje
+---
+
+# Interacting With Files
+
+Directus exposes services that allow you to perform operations programmatically on your files.
+
+## FilesService
```
Comment (phazonoverload) [inline docs/extensions/services/interacting-with-files.md:10]: I would remove this for consistency with ItemsService. I also think it's slightly redundant.

File: docs/extensions/services/interacting-with-files.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+The `FilesService` provides access to upload, import and perform other CRUD operations on files in Directus Files.
```
Comment (phazonoverload) [inline docs/extensions/services/interacting-with-files.md:12]:
```suggestion
The `FilesService` provides access to upload, import, and perform other operations on files in Directus Files.
```

Comment (phazonoverload) [reply docs/extensions/services/interacting-with-files.md:12]: We love an oxford comma

File: docs/extensions/services/interacting-with-files.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+    res.json(data);
+  });
+```
+
+::: tip Explore FilesService In-depth
+
+Check out the full list [here](https://github.com/directus/directus/blob/main/api/src/services/files.ts).
+
+:::
```
Comment (phazonoverload) [inline docs/extensions/services/interacting-with-files.md:63]: Please apply the same changes to this file as the ItemsService page: 

1. Rename based on nav suggestions
2. Add a generic 'this is how you initialize' example before the first method. 
3. Fix up indentation. 
4. End of file make sure link contains explainer.

File: docs/extensions/services/interacting-with-files.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    const service = new FilesService({ schema });
+    const assetKey = await service.importOne(req.body.url);
+
+    const data = await service.readOne(assetKey);
+    res.locals['payload'] = { data: data || null };
+
+    res.json(data);
+  });
+```
+
```
Comment (phazonoverload) [inline docs/extensions/services/interacting-with-files.md:28]: Upload a file is missing

File: docs/extensions/services/working-with-users.md [hunk 13]
```diff
@@ -0,0 +1,104 @@
+---
+description: Learn about the required services when working with users and granting them access control.
+contributors: Esther Agbaje
+---
+
+# Working with Users
+
+Directus provides a comprehensive set of services that facilitate seamless collaboration with users while enforcing
+access control to specific roles. They include:
```
Comment (phazonoverload) [inline docs/extensions/services/working-with-users.md:9]: This reads very marketing to me. I would make it more straightforward. Something like: 

> There are various services related to users:

File: docs/extensions/services/working-with-users.md [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+- `UsersService`: Extends the `ItemsService` and is used to manage user accounts and perform CRUD operations on user
+  profiles
+- `RolesServices`: Extends the `ItemsService` and enables you to assign roles to users.
+- `PermissionsServices`: Extends the `ItemsService` and allows you to manage the permissions associated with each role.
```
Comment (phazonoverload) [inline docs/extensions/services/working-with-users.md:14]: Remove 'Extends the ItemsService' at the start and get to what they do. Then just add it as a sentence under the list. 'All three of these services extend the Items service so are used in a very similar way' or something like that linking to the ItemsService page.

File: docs/extensions/services/working-with-users.md [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
```
Comment (phazonoverload) [inline docs/extensions/services/working-with-users.md:15]: Create a user?

File: docs/extensions/services/interacting-with-files.md [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    const service = new FilesService({ schema });
+    const { id } = req.params;
+
+    const data = await service.readOne(id);
+    res.locals['payload'] = { data: data || null };
+
+    res.json(data);
+  });
+```
+
```
Comment (phazonoverload) [inline docs/extensions/services/interacting-with-files.md:43]: Update missing

File: docs/extensions/services/working-with-users.md [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+    const service = new UsersService({ schema });
+    const { email } = req.params;
+
+    const data = await service.getUserByEmail(email);
+    res.locals['payload'] = { data: data || null };
+
+    res.json(data);
+  });
+```
+
```
Comment (phazonoverload) [inline docs/extensions/services/working-with-users.md:30]: Update, Delete. Do the full CRUD for users.

File: docs/extensions/services/working-with-users.md [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+### Create and Assign User to a Role
+
+```js
+ router.post('/users/:role', async (req, res) => {
+    const schema = await getSchema();
+    const service = new UsersService({ schema });
+    const roleService = new RolesService({ schema });
```
Comment (phazonoverload) [inline docs/extensions/services/working-with-users.md:49]: name them userService and roleService for clarity

File: docs/extensions/services/working-with-users.md [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+ router.post('/roles', async (req, res) => {
+    const schema = await getSchema();
+    const service = new RolesService({ schema });
+    const data = await service.createOne(req.body);
+    res.locals['payload'] = { data: data || null };
+    res.json(data);
+  });
+```
+
+### Create and Assign User to a Role
```
Comment (phazonoverload) [inline docs/extensions/services/working-with-users.md:43]: Create a User and Assign a Role

Comment (phazonoverload) [reply docs/extensions/services/working-with-users.md:43]: More clear. I thought initially this created a role then added it to an existing user.

File: docs/extensions/services/working-with-users.md [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+    res.json(data);
+  });
+```
+
+### Assign Permission to a Role
+
+```js
+router.post('/permissions/:role', async (req, res) => {
+    const schema = await getSchema();
+    const service = new PermissionsService({ schema });
```
Comment (phazonoverload) [inline docs/extensions/services/working-with-users.md:76]: rename to permissionService for clarity

File: docs/extensions/services/working-with-users.md [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+    const roleService = new RolesService({ schema });
+
+    const { role } = req.params;
+    const roles = await roleService.readByQuery({
+      fields: ['*'],
+    });
+
+    const foundRole = roles.find(
+      (item: any) => item.name.toLowerCase() === role.toLowerCase()
+    );
```
Comment (phazonoverload) [inline docs/extensions/services/working-with-users.md:58]: Can we neaten this up and assume the role name is being passed in lowercase and matches. Update the query to match where `name == role`

File: docs/extensions/services/working-with-users.md [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+      role: foundRole.id,
+    });
+
+    res.locals['payload'] = { data: data || null };
+
+    res.json(data);
+  });
+```
+
+### Assign Permission to a Role
```
Comment (phazonoverload) [inline docs/extensions/services/working-with-users.md:71]: Add this just below 'create a role'. Feels like the natural next one.

File: docs/extensions/services/working-with-users.md (hunk 15 above)
Comment (phazonoverload) [reply docs/extensions/services/working-with-users.md:15]: You cover it below as part of create user and assign role, but I think it should be standalone too.

File: docs/extensions/services/working-with-users.md (hunk 21 above)
Comment (phazonoverload) [reply docs/extensions/services/working-with-users.md:58]: Same for others

File: docs/extensions/services/working-with-users.md [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+    const roles = await roleService.readByQuery({
+      fields: ['*'],
+    });
+
+    const foundRole = roles.find(
+      (item: any) => item.name.toLowerCase() === role.toLowerCase()
+    );
+
+    const data = await service.createOne({
+      ...req.body,
```
Comment (phazonoverload) [inline docs/extensions/services/working-with-users.md:61]: req.body is fine but you need to give an example of what's in the body. In the 'create a user example' please just hardcode the values. A developer reading this should understand they need to provide dynamic values.

File: docs/extensions/services/working-with-users.md [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+
+    res.json(data);
+  });
+```
+
+::: tip Explore Services In-depth
+
+Check out the full list [here](https://github.com/directus/directus/blob/main/api/src/services).
+
+:::
```
Comment (phazonoverload) [inline docs/extensions/services/working-with-users.md:104]: Please apply the same changes to this file as the ItemsService page: 

1. Rename based on nav suggestions
2. Add a generic 'this is how you initialize' example before the first method. I'll leave it to you to figure out if it needs to be three examples, one where you say after 'it works the same for all three', or some other permutation. 
3. Fix up indentation. 
4. End of file make sure link contains explainer.

File: docs/extensions/services/working-with-users.md (hunk 22 above)
Comment (phazonoverload) [reply docs/extensions/services/working-with-users.md:71]: This one is also hard to follow. Please hardcode values instead of req.body so we can see what the permissions service expects.

File: docs/extensions/services/authoring-database.md [hunk 25]
```diff
@@ -0,0 +1,187 @@
+---
+description: Learn about the Services for authoring the database in Directus and how to utilize them when building extensions.
+contributors: Esther Agbaje
+---
+
+# Authoring the Database
+
+Directus provides services such as `CollectionsService`, `FieldsService` and `RelationsService` to programmatically
+configure and modify the database schema.
```
Comment (phazonoverload) [inline docs/extensions/services/authoring-database.md:9]:
```suggestion
Use the `CollectionsService`, `FieldsService` and `RelationsService` to
configure and modify the data model.
```

File: docs/extensions/services/authoring-database.md [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+
+## CollectionsService
+
+The `CollectionsService` is responsible for managing and manipulating data in collections. Utilize it to create CRUD
+endpoints to collections.
```
Comment (phazonoverload) [inline docs/extensions/services/authoring-database.md:14]: Create CRUD is like Pin Number. 'Perform' might be better.

File: docs/extensions/services/authoring-database.md [hunk 27]
```diff
@@ -0,0 +1,187 @@
...
+
+    res.json(data);
+  });
+```
+
+::: tip Explore Services In-depth
+
+Check out the full list [here](https://github.com/directus/directus/blob/main/api/src/services).
+
+:::
```
Comment (phazonoverload) [inline docs/extensions/services/authoring-database.md:187]: Please apply the same changes to this file as the ItemsService page: 

1. Rename based on nav suggestions
2. Add a generic 'this is how you initialize' example before the first method. in each section. 
3. Fix up indentation. 
4. End of file make sure link contains explainer.

**Review by phazonoverload — CHANGES_REQUESTED** [review]
Good start - a few changes. I did the review by looking at the config, intro, items, and then the others. I reference comments I did in that order so I'd encourage you to go through the review in that order. 

The only other global comment I would make is that in an effort to provide 'real world' examples, you've obfuscated important information. I'd suggest every single example is for the root `/` route handler and you hardcode values. Don't use `req.id` but instead `'item_id'`. Same goes for `req.body` and `{ property: value, property2: value2 }`, and so on.

When you're ready please re-request a review.

File: docs/extensions/services/introduction.md [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+```js
+export default defineEndpoint((router, context) => {
+  const { services, getSchema } = context;
+  const { ItemsService } = services;
+
+})
+```
+
+This allows each endpoint implementations to initialize an instance of that service and leverage its methods to perform
+CRUD operations.
```
[the remark this answers is not in the document]
Comment (estheragbaje) [reply docs/extensions/services/introduction.md:30]: rephrased this

File: docs/extensions/services/interacting-with-files.md (hunk 12 above)
Comment (estheragbaje) [reply docs/extensions/services/interacting-with-files.md]: Encountered a couple of errors while trying to implement the endpoint to upload a file- why I left it out. Also, imo, there doesn't seem to be a strong use case for uploading files via the API as most users will already have the url format of the file, which is what importing a file does.

Comment (phazonoverload) [reply docs/extensions/services/interacting-with-files.md]: Don't agree - what if I want to build a panel that has a file upload functionality? Let's try and get this resolved - happy to pair with you 😄

Comment (br41nslug) [reply docs/extensions/services/interacting-with-files.md]: Because "receiving a file" on a custom endpoint is very tricky I'd recommend having an endpoint example that takes a base64 image string and uploads that to the file library.

Comment (phazonoverload) [reply docs/extensions/services/interacting-with-files.md]: Ok but that's still a viable approach to document, no? Or do you think it reaches outside of the common asks people have.

File: docs/extensions/services/accessing-database.md (hunk 7 above)
Comment (estheragbaje) [reply docs/extensions/services/accessing-database.md]: `getSchema()` can be called at the top level. Moved it there.

Accountabilility doesn't show up as a valid type. See screenshot
Screenshot 2023-11-20 at 19 54 22

When the type is fixed, I'll circle back and update it accordingly.

**Review by phazonoverload — APPROVED** [review]

**Comment by phazonoverload** [comment]
Very happy with this. Good set of references.
